torch/nested/_internal/nested_tensor.py

```diff
@@ -3,23 +3,224 @@
+    #  j0_0 ------> vec0 -----> equiv_set0 ---> sum_vec: Optional[SymInt]
+    #  j0_1 ------^         /-^            \--> vec_{cuda,cpu}: Optional[Tensor]
+    #                      /
+    #               vec1--/               /---- weak vec2, weak j1_0
+    #                                    | /--> id: int = 1
+    #  j1_0 ------> vec2 -----> equiv_set1 ---> ...
+    #
+    # Details on canonical vec and nested int lifetimes:
+    # --------------------------------------------------
+    # Throughout the lifetime of a equiv set, the canonical vec may change.
+    # If we detect that the canonical vec is no longer alive, the next time
+    # someone asks us for a nested int for a vec in that equiv set we will make
+    # that vec the canonical vec. The canonical vec may be on cpu or cuda.
+    #
+    # If canonical vec and nested int are alive, nested int's vec must be the
+    # canonical vec. This implies that if canonical is not alive, then the
+    # nested int is also not alive. It is possible for canonical to be alive,
+    # but for nested int to be dead. This happens when the nested int is only
+    # alive in cpp, but its corresponding python object has died.
+    #
+    # When you have tracing subclass tensors as vec
+    # ---------------------------------------------
+    # During tracing, we may be dealing with vec which are subclasses used
+    # for tracing, e.g. FakeTensor and FunctionalTensor. Though these tensors
+    # don't actually need equiv set for the purpose of checking equality, we
+    # the concepts of equiv sets are still useful so that we can reuse the logic
+    # in eager to access the cpu and cuda variants from one another. Note that
+    # that since each equiv set has its own cache, FakeTensor must be in a
+    # different equiv set from real tensor.
+    #
+    def __init__(self):
+        self._equiv_set_counter = 0
+        self._equiv_sets = WeakTensorKeyDictionary()
+        self._version_counters = WeakTensorKeyDictionary()
+
+    def contains_vec(self, vec):
+        return vec in self._equiv_sets
 
-def get_tensor_symint(tensor, *, coeff=1):
-    global _tensor_id_counter
-    tensor_symint = _tensor_symint_registry.get(tensor)
-    if tensor_symint is None:
-        tensor_symint = torch._C._get_nested_int(_tensor_id_counter, coeff)
-        _tensor_id_counter += 1
-        _tensor_symint_registry[tensor] = tensor_symint
-    return tensor_symint
+    def assert_contains_vec(self, vec):
+        assert self.contains_vec(vec), (
+            "Expected vec to have been registered. "
+        )
 
+    def check_version_counter(self, vec):
+        # Check that vec has not been mutated
+        assert self._version_counters[vec] == vec._version, (
+            "Detected that vec has been mutated. This is not allowed. "
+        )
+
+    def maybe_create(self, vec, *, ctor_fn=None, equiv_set_from=None, has_coeff=False):
+        # Given vec, return an associated nested int.
+        #
```

Comment:
Add some discussion about what the user can expect from this API, e.g. the same vec implies nested int with same equiv id, but not necessarily the same instance. If the vec is later mutated, using it with the registry will error.